File: TitanEngine/SDK/Samples/Tools/PEValidate/PEValidate/PEValidate.cpp
// PEValidate.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"
#include "windows.h"
#include "SDK\SDK.h"

char* GetOveralEval(int EvalValue){

	if(EvalValue == UE_RESULT_FILE_OK){
		return("File is OK!");
	}else if(EvalValue == UE_RESULT_FILE_INVALID_BUT_FIXABLE){
		return("File is invalid but fixable!");
	}else if(EvalValue == UE_RESULT_FILE_INVALID_AND_NON_FIXABLE){
		return("File is invalid and non fixable!");
	}else if(EvalValue == UE_RESULT_FILE_INVALID_FORMAT){
		return("File is invalid format!");
	}else{
		return("INVALID STRUCTURE VALUE");
	}
}

char* GetFieldEval(int EvalValue){

	if(EvalValue == UE_FIELD_OK){
		return("OK");
	}else if(EvalValue == UE_FIELD_BROKEN_NON_FIXABLE){
		return("FIXABLE");
	}else if(EvalValue == UE_FIELD_BROKEN_NON_CRITICAL){
		return("NON CRITICAL");
	}else if(EvalValue == UE_FIELD_BROKEN_FIXABLE_FOR_STATIC_USE){
		return("FIXABLE:STATIC");
	}else if(EvalValue == UE_FIELD_BROKEN_BUT_CAN_BE_EMULATED){
		return("BROKEN:EMULATE");
	}else if(EvalValue == UE_FILED_FIXABLE_NON_CRITICAL){
		return("FIXABLE:OPTIONAL");
	}else if(EvalValue == UE_FILED_FIXABLE_CRITICAL){
		return("FIXABLE:CRITICAL");
	}else if(EvalValue == UE_FIELD_NOT_PRESET){
		return("NOT PRESENT");
	}else if(EvalValue == UE_FIELD_NOT_PRESET_WARNING){
		return("NOT PRESENT:WARNING");
	}else{
		return("INVALID STRUCTURE VALUE");
	}
}

char* GetBooleanEval(int EvalValue){

	if(EvalValue == 1){
		return("True");
	}else{
		return("False");
	}
}

int _tmain(int argc, _TCHAR* argv[]){

	FILE_STATUS_INFO fileStatus = {};

	printf("PEValidator 1.0\r\nReversingLabs Corporation - www.reversinglabs.com\r\n\r\n");
	if(argc > 1){
		if(IsPE32FileValidExW(argv[1], UE_DEPTH_DEEP, &fileStatus)){
			printf(" [+] Selected file is valid!\r\n");
		}else{
			printf(" [-] Selected file is NOT valid!\r\n");
			printf(" [-] Selected file OveralEvaluation: %s\r\n", GetOveralEval(fileStatus.OveralEvaluation));
			printf("     [x] FileIsDLL: %s\r\n", GetBooleanEval(fileStatus.FileIsDLL));
			printf("     [x] FileIsConsole: %s\r\n", GetBooleanEval(fileStatus.FileIsConsole));
			printf("     [x] FileIs64Bit: %s\r\n", GetBooleanEval(fileStatus.FileIs64Bit));
			printf("     [x] EvaluationTerminatedByException: %s\r\n", GetBooleanEval(fileStatus.EvaluationTerminatedByException));
			printf("     [x] MissingDependencies: %s\r\n", GetBooleanEval(fileStatus.MissingDependencies));
			printf("     [x] MissingDeclaredAPIs: %s\r\n", GetBooleanEval(fileStatus.MissingDeclaredAPIs));
			printf("     [-] File details:\r\n");
			printf("         [x] SignatureMZ: %s\r\n", GetFieldEval(fileStatus.SignatureMZ));
			printf("         [x] SignaturePE: %s\r\n", GetFieldEval(fileStatus.SignaturePE));
			printf("         [x] EntryPoint: %s\r\n", GetFieldEval(fileStatus.EntryPoint));
			printf("         [x] ImageBase: %s\r\n", GetFieldEval(fileStatus.ImageBase));
			printf("         [x] SizeOfImage: %s\r\n", GetFieldEval(fileStatus.SizeOfImage));
			printf("         [x] FileAlignment: %s\r\n", GetFieldEval(fileStatus.FileAlignment));
			printf("         [x] SectionAlignment: %s\r\n", GetFieldEval(fileStatus.SectionAlignment));
			printf("         [x] ExportTable: %s\r\n", GetFieldEval(fileStatus.ExportTable));
			printf("         [x] RelocationTable: %s\r\n", GetFieldEval(fileStatus.RelocationTable));
			printf("         [x] ImportTable: %s\r\n", GetFieldEval(fileStatus.ImportTable));
			printf("         [x] ImportTableSection: %s\r\n", GetFieldEval(fileStatus.ImportTableSection));
			printf("         [x] ImportTableData: %s\r\n", GetFieldEval(fileStatus.ImportTableData));
			printf("         [x] IATTable: %s\r\n", GetFieldEval(fileStatus.IATTable));
			printf("         [x] TLSTable: %s\r\n", GetFieldEval(fileStatus.TLSTable));
			printf("         [x] LoadConfigTable: %s\r\n", GetFieldEval(fileStatus.LoadConfigTable));
			printf("         [x] BoundImportTable: %s\r\n", GetFieldEval(fileStatus.BoundImportTable));
			printf("         [x] COMHeaderTable: %s\r\n", GetFieldEval(fileStatus.COMHeaderTable));
			printf("         [x] ResourceTable: %s\r\n", GetFieldEval(fileStatus.ResourceTable));
			printf("         [x] ResourceData: %s\r\n", GetFieldEval(fileStatus.ResourceData));
			printf("         [x] SectionTable: %s\r\n", GetFieldEval(fileStatus.SectionTable));
			printf("\r\n");
		}
	}else{
		printf(" Usage: PEValidate inputFileName.ext\r\n");
	}
	return 0;
}

File: TitanEngine/SDK/Samples/Scripts/LUA/Tools/PEValidate.lua
require 'SDK'

function GetOveralEval(EvalValue)

	if EvalValue == UE_RESULT_FILE_OK then
		return "File is OK!"
	elseif EvalValue == UE_RESULT_FILE_INVALID_BUT_FIXABLE then
		return "File is invalid but fixable!"
	elseif EvalValue == UE_RESULT_FILE_INVALID_AND_NON_FIXABLE then
		return "File is invalid and non fixable!"
	elseif EvalValue == E_RESULT_FILE_INVALID_FORMAT then
		return "File is invalid format!"
	else
		return "INVALID STRUCTURE VALUE"
	end
end

function GetFieldEval(EvalValue)

	if EvalValue == UE_FIELD_OK then
		return "OK"
	elseif EvalValue == UE_FIELD_BROKEN_NON_FIXABLE then
		return "FIXABLE"
	elseif EvalValue == UE_FIELD_BROKEN_NON_CRITICAL then
		return "NON CRITICAL"
	elseif EvalValue == UE_FIELD_BROKEN_FIXABLE_FOR_STATIC_USE then
		return "FIXABLE:STATIC"
	elseif EvalValue == UE_FIELD_BROKEN_BUT_CAN_BE_EMULATED then
		return "BROKEN:EMULATE"
	elseif EvalValue == UE_FILED_FIXABLE_NON_CRITICAL then
		return "FIXABLE:OPTIONAL"
	elseif EvalValue == UE_FILED_FIXABLE_CRITICAL then
		return "FIXABLE:CRITICAL"
	elseif EvalValue == UE_FIELD_NOT_PRESET then
		return "NOT PRESENT"
	elseif EvalValue == UE_FIELD_NOT_PRESET_WARNING then
		return "NOT PRESENT:WARNING"
	else
		return "INVALID STRUCTURE VALUE"
	end
end

function GetBooleanEval(EvalValue)

	if EvalValue == 1 then
		return "True"
	else
		return "False"
	end
end


fileStatus = FILE_STATUS_INFO:new()

fileInfo = TE_IsPE32FileValidEx(arg[1], UE_DEPTH_DEEP, fileStatus())
if not(fileInfo == nil) then
	print(" [+] Selected file is valid!")
else
	print(" [-] Selected file is NOT valid!");
	print(" [-] Selected file OveralEvaluation: " .. GetOveralEval(fileStatus.OveralEvaluation))
	print("     [x] FileIsDLL: " .. GetBooleanEval(fileStatus.FileIsDLL))
	print("     [x] FileIsConsole: " .. GetBooleanEval(fileStatus.FileIsConsole))
	print("     [x] FileIs64Bit: " .. GetBooleanEval(fileStatus.FileIs64Bit))
	print("     [x] EvaluationTerminatedByException: " .. GetBooleanEval(fileStatus.EvaluationTerminatedByException))
	print("     [x] MissingDependencies: " .. GetBooleanEval(fileStatus.MissingDependencies))
	print("     [x] MissingDeclaredAPIs: " .. GetBooleanEval(fileStatus.MissingDeclaredAPIs))
	print("     [-] File details: ")
	print("         [x] SignatureMZ: " .. GetFieldEval(fileStatus.SignatureMZ))
	print("         [x] SignaturePE: " .. GetFieldEval(fileStatus.SignaturePE))
	print("         [x] EntryPoint: " .. GetFieldEval(fileStatus.EntryPoint))
	print("         [x] ImageBase: " .. GetFieldEval(fileStatus.ImageBase))
	print("         [x] SizeOfImage: " .. GetFieldEval(fileStatus.SizeOfImage))
	print("         [x] FileAlignment: " .. GetFieldEval(fileStatus.FileAlignment))
	print("         [x] SectionAlignment: " .. GetFieldEval(fileStatus.SectionAlignment))
	print("         [x] ExportTable: " .. GetFieldEval(fileStatus.ExportTable))
	print("         [x] RelocationTable: " .. GetFieldEval(fileStatus.RelocationTable))
	print("         [x] ImportTable: " .. GetFieldEval(fileStatus.ImportTable))
	print("         [x] ImportTableSection: " .. GetFieldEval(fileStatus.ImportTableSection))
	print("         [x] ImportTableData: " .. GetFieldEval(fileStatus.ImportTableData))
	print("         [x] IATTable: " .. GetFieldEval(fileStatus.IATTable))
	print("         [x] TLSTable: " .. GetFieldEval(fileStatus.TLSTable))
	print("         [x] LoadConfigTable: " .. GetFieldEval(fileStatus.LoadConfigTable))
	print("         [x] BoundImportTable: " .. GetFieldEval(fileStatus.BoundImportTable))
	print("         [x] COMHeaderTable: " .. GetFieldEval(fileStatus.COMHeaderTable))
	print("         [x] ResourceTable: " .. GetFieldEval(fileStatus.ResourceTable))
	print("         [x] ResourceData: " .. GetFieldEval(fileStatus.ResourceData))
	print("         [x] SectionTable: " .. GetFieldEval(fileStatus.SectionTable))
end
